### PR TITLE
Add Execution Platform Version That Would Fix Lint Error

### PR DESCRIPTION
**Pull request by philippfromme**
The `allowedVersion` property will be used by [`camunda/linting`](https://github.com/camunda/linting) to provide a hint to the user.

# Changes

* rules that need configuration (e.g. `element-type`) are now configured by specifying execution platform version
* `element-type` rule _only_ reports `ELEMENT_TYPE_NOT_ALLOWED` errors
* `timer` rule reports new error types `EXPRESSION_REQUIRED` and `EXPRESSION_VALUE_NOT_ALLOWED`
* `report.error` was renamed to `report.data` as it is meant to be used for any additional data
* `ERROR_TYPES` were namespaced to avoid mistaking reports with `report.data.type` for Camunda-specific reports
* `FEEL_INVALID` error type renamed to `FEEL_EXPRESSION_INVALID`
* `feel` and `timer` rule handle `bpmn:Expression` elements
* integration tests were rewritten for maintainability

This pull request includes a number of breaking changes. For users of `@camunda/linting` there are no breaking changes unless they work with details of the reports.

---

Related to https://github.com/camunda/camunda-modeler/issues/3148

### CODE REVIEW

**Review by nikku — COMMENTED** [review]
I like this change overall. Two things we could consider:

* Add documentation "contract" so integrators can do something meaningful with the meta-data
* Consider namespacing this to prevent (future) clashes with other linter rules: Call it `camundaCompatAllowedVersion` over `allowedVersion`

Alternative to consider:

* Explicitly expose meta-data with the error: `error = { ..., meta }`
* `meta = { supportedCamundaPlatform, category: 'execution', ... }`

This could prepare us for modeling only + categories coming and clearly separate errors from "useful meta-data".


Alternative to consider: 

* Shall this meta-data be attached to the error, or to the rule
* If attached to the rule, then rules must be fine-grained "enough". On the positive side that means that meta-data can be attached to the error automatically => We have a very clear programming model.

**Comment by philippfromme** [comment]
> Alternative to consider:
> 
> * Shall this meta-data be attached to the error, or to the rule
> * If attached to the rule, then rules must be fine-grained "enough". On the positive side that means that meta-data can be attached to the error automatically => We have a very clear programming model.

This would mean separate `element-type` rules for all execution platform versions. Yes, we could attach the metadata to the rule itself, but on the other hand, we couldn't make use of configuration anymore.

In any case, let's discuss this!

**Comment by nikku** [comment]
@philippfromme Let's conduct a quick session next week to discuss this. I want to make sure that we don't do double the work once we introduce categories ("implement" / "model") or other meta-data in the future.

**Comment by philippfromme** [comment]
@marstamm This is ready to be reviewed. We'll rebase https://github.com/camunda/bpmnlint-plugin-camunda-compat/pull/52 afterwards.

**Comment by philippfromme** [comment]
> @marstamm This is ready to be reviewed. We'll rebase #52 afterwards.

Since this pull request has already been merged I rebased.

**Review by marstamm — COMMENTED** [review]
Quite a refactoring!
If I want to try it out, is this the correct branch for integration into `@camunda/linting`? https://github.com/camunda/linting/tree/allowed-version

As we merged https://github.com/camunda/bpmnlint-plugin-camunda-compat/pull/52 before this one, do we need to change `error` to `data` there as well?
https://github.dev/camunda/bpmnlint-plugin-camunda-compat/blob/f272518064fd23d2c3b0db5de6bc5a4109536137/rules/collapsed-subprocess.js#L18

File: rules/called-decision-or-task-definition/index.js [hunk 1]
```diff
@@ -16,14 +17,11 @@ const { reportErrors } = require('../utils/reporter');
...
-    elementsCalledDecision = [],
-    elementsTaskDefinition = []
-  } = config;
-
+module.exports = function({ version }) {
   function check(node, reporter) {
-    if (!isAny(node, elementsCalledDecision) && !isAny(node, elementsTaskDefinition)) {
+    if (
+      (!config.calledDecision[ node.$type ] || !greaterOrEqual(version, config.calledDecision[ node.$type ]))
+      && (!config.taskDefinition[ node.$type ] || !greaterOrEqual(version, config.taskDefinition[ node.$type ]))) {
```
[the remark this answers is not in the document]
Comment (philippfromme) [reply rules/called-decision-or-task-definition/index.js:24]: Let me double-check that.

File: rules/timer/config.js [hunk 2]
```diff
@@ -1,17 +1,4 @@
...
-  camundaCloud10: {
-    formats: camundaCloud10Formats
-  },
-  camundaCloud81: {
-    formats: camundaCloud81Formats
-  }
-};
+  'cron': '8.1',
+  'iso8601': '1.0'
+};
```
[the remark this answers is not in the document]
Comment (philippfromme) [reply rules/timer/config.js:4]: 🚀

**Comment by philippfromme** [comment]
> If I want to try it out, is this the correct branch for integration into `@camunda/linting`? https://github.com/camunda/linting/tree/allowed-version

Yes, this is the corresponding branch. Due to the breaking changes other @camunda/linting pull requests have to be adjusted slightly.

File: rules/called-decision-or-task-definition/index.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (philippfromme) [reply rules/called-decision-or-task-definition/index.js:24]: I think this is fine, we could have used the strict checks before. For now, we are explicit about the element types. We ignore inheritance.

[the remark this answers is not in the document]
Comment (marstamm) [reply rules/called-decision-or-task-definition/index.js:24]: Agreed, we are fine here

**Comment by marstamm** [comment]
Integration works fine as well. I made the changes to the subprocess rule in [4e474bf](https://github.com/camunda/bpmnlint-plugin-camunda-compat/pull/50/commits/4e474bf0c2cfc385170c49bf599c8343a57b2b8e)

**Comment by marstamm** [comment]
Do we want to squash the commits before merging?

**Comment by philippfromme** [comment]
> Do we want to squash the commits before merging?

You mean regarding the last commit you added? I can do that.

**Review by marstamm — APPROVED** [review]

**Pull request merged**